[lld][AArch64][Build Attributes] Add support for AArch64 Build Attributes

sivan-shani opened the pull request:
This patch enables lld to read AArch64 Build Attributes and convert them into GNU Properties.

Changes:
- Parses AArch64 Build Attributes from input object files.
- Converts known attributes into corresponding GNU Properties.
- Merges attributes when linking multiple objects.

sivan-shani committed:
[lld] Refactor storage of PAuth ABI core info

Previously, the AArch64 PAuth ABI core values were stored as an ArrayRef<uint8_t>, introducing unnecessary indirection.
This patch replaces the ArrayRef with two explicit uint64_t fields: aarch64PauthAbiPlatform and aarch64PauthAbiVersion. This simplifies the representation and improves readability.
No functional change intended, aside from improved error messages.

sivan-shani committed:
[lld][AArch64][Build Attributes] Add support for AArch64 Build Attributes

This patch enables lld to read AArch64 Build Attributes and convert them into GNU Properties.

Changes:
- Parses AArch64 Build Attributes from input object files.
- Converts known attributes into corresponding GNU Properties.
- Merges attributes when linking multiple objects.

smithp35 reviewed (commented):
Just one quick comment on the functional side (mapping PAuth BuildAttributes 0, 1 to an explicit GNU properties of 0, 0).

Will keep looking to see if I can spot anything else.

sivan-shani committed:
Map (TagPlatform, TagSchema) = (0, 1) to explicit property of (0, 0)

According to the Arm BuildAttributes specification, a (TagPlatform, TagSchema)
of (0, 1) corresponds to an explicit property with platform = 0 and version = 0 ('Invalid').

Update LLD to handle this case by mapping it to aarch64PauthAbiCoreInfo = {0, 0}.
Update comment to clarify default behavior and mapping.

Spec reference:
https://github.com/ARM-software/abi-aa/pull/230/files#r1030

sivan-shani commented on lld/ELF/InputFiles.cpp:
```diff
@@ -539,6 +540,42 @@ uint32_t ObjFile<ELFT>::getSectionIndex(const Elf_Sym &sym) const {
+                  "properties and AArch64 build attributes sections with "
+                  "different And Features data";
+  } else {
+    // Write missing data
```
Handled
